### DISCUSSION TITLE
Revalidating data from websockets

### DISCUSSION THREAD

**Discussion by Xelson** [remix-run/remix#3849] [Q&A]
Let's suppose I have an notifications model:
```js
export interface Notification {
	id: number,
	title: string,
	body: string,
	recipient: string | null,
	type: NotificationType,
	created_at: string
}

export function useNotifications() : Notification[] {
	const notifications = useRouteData('routes/panel').notifications;
	return notifications;
}
```

I get access to notifications data through a wrapper of `useMatches` hook:
```js
export function useRouteData<Data>(routeId: string): Data | any {
  return useMatches().find((match) => match.id === routeId)?.data as
    | Data
    | any;
}
```
Now we can get up to date data from the `/panel` route via react hook
```js
export function RenderNotifications() {
	const notifications = useNotifications();

	return (
		<div>
			{notifications.map(notification => {
				<NotificationCard 
					key={notification.id} 
					notification={notification} 
				/>
			})}
		</div>
	);
}
```
This list will always be redrawn when data is revalidated by submitting a form with post action. 
But what should I do from the point of view of optimistic UI if updates come via `websocket`? The problem is if I allow the user to initiate dummy post action for data revalidating for many users then I will provoke a kind of ddos attack on my own service because users will make a request at the same time as accepting a socket message. 
How could I update the data by revalidating the data from the web sockets, initiating a rerendeing for the `useMatches` hook?

**Comment by kiliman** [comment]
If your websocket is sending you updated data, I don't think I would revalidate the same data. Perhaps `unstable_shouldReload` on that route. At that point, I would maintain local React state that is updated from the socket message. And let React re-render. You should pick one or the other.

Another option is to have the websocket simply notify you have a change and your client can revalidate. That way the standard remix data flow will trigger rendering.

**Reply by Xelson** [reply]
I repeat myself that sometimes it's just not rational to notify server that i have a change and my client can revalidate because it can happen as a flash of requests to the server, which is not very expedient. This is reason why i want to make local refreshes of data — the changes are already known and there is no need to ask the server about it.
I assumed it was a matter of maintaining the local state but I would like a more concise solution, it just looks like Remix doesn't have an architectural solution for local refreshes, and this is very sad.
It seems to me that the solution would be to edit `data` property directly in route match so in case of default data update from remix data flow we will be able to revalidate the data directly from the server, without creating the need to create a local state everywhere where local data refresh is needed for components

**Reply by kiliman** [reply]
You know your use case better than me. I'm just saying that it's best to have a single source of truth for the data that is being rendered. In this case I recommend React `useState`.

## Initial Data
1) useLoaderData() 
2) or useFetcher() after initial render
3) put this in React state and render your list

## Add new data local from Form
1) Submit to action (but don't revalidate)
2) Use optimistic UI and add the new data directly to local state to render

## Web Socket update
1) New updates that come from server
2) Add to local state and render

If user ever refreshes, they'll get Initial Data again which should include data that use added or sent via sockets, and the cycle repeats.

This is a similar technique used for infinite scrolling... keep a subset of the data (visible portion + extra) in local state.... initially filled from useLoaderData. Then useFetcher to get more data as use scrolls... update local state (possible dropping some data to keep the working set manageable).

**Reply by kiliman** [reply]
Note, if Web Socket notifies on current user changes, then you'll need to filter those out by ensuring you don't add duplicates to your local state... possibly check for existing id.
